src/layers/invertible_layer_irim.jl

```diff
@@ -79,92 +87,120 @@ CouplingLayerIRIM3D(args...;kw...) = CouplingLayerIRIM(args...; kw..., ndims=3)
+
+        tensor_cat!(Y_, Y1_, Y2_)
+        X = L.C[j].inverse(Y_)
+    end
     
-    return Y
+    return X
 end
 
 # 2D Inverse pass: Input Y, Output X
 function inverse(Y::AbstractArray{T, N}, L::CouplingLayerIRIM; save=false) where {T, N}
-    Y_ = L.C.forward(Y)
-    Y1_, Y2_ = tensor_split(Y_)
 
-    X1_ = Y1_
-    X2_ = Y2_ - L.RB.forward(Y1_)
+    # Init tensors to avoid reallocation
+    X_ = similar(Y)
+
+    num_downsamp = length(L.C)
+    for j=num_downsamp:-1:1
+        Y_ = L.C[j].forward(Y)
+        Y1_, Y2_ = tensor_split(Y_)
 
-    X_ = tensor_cat(X1_, X2_)
-    X = L.C.inverse(X_)
+        X1_ = Y1_
+        X2_ = Y2_ - L.RB[j].forward(Y1_)
 
-    if save == false
-        return X
-    else
-        return X, X_, Y1_
+        tensor_cat!(X_, X1_, X2_)
+        Y = L.C[j].inverse(X_)
     end
+
+    return Y
 end
 
 # 2D Backward pass: Input (ΔY, Y), Output (ΔX, X)
 function backward(ΔY::AbstractArray{T, N}, Y::AbstractArray{T, N}, L::CouplingLayerIRIM; set_grad::Bool=true) where {T, N}
 
-    # Recompute forward state
-    k = Int(L.C.k/2)
-    X, X_, Y1_ = inverse(Y, L; save=true)
-
-    # Backpropagate residual
-    if set_grad
-        ΔY_ = L.C.forward((ΔY, Y))[1]
-    else
-        ΔY_, Δθ_C1 = L.C.forward((ΔY, Y); set_grad=set_grad)[1:2]
-    end
-    ΔYl_, ΔYr_ = tensor_split(ΔY_)
-    if set_grad
-        ΔY1_ = L.RB.backward(ΔYr_, Y1_) + ΔYl_
-    else
-        ΔY1_, Δθ_RB = L.RB.backward(ΔYr_, Y1_; set_grad=set_grad)
-        ΔY1_ = ΔY1_ + ΔYl_
-    end
-    
-    ΔX_ = tensor_cat(ΔY1_, ΔYr_)
-    if set_grad
-        ΔX = L.C.inverse((ΔX_, X_))[1]
-    else
-        ΔX, Δθ_C2 = L.C.inverse((ΔX_, X_); set_grad=set_grad)[1:2]
+    # Initialize layer parameters
+    !set_grad && (p1 = Array{Parameter, 1}(undef, 0))
+    !set_grad && (p2 = Array{Parameter, 1}(undef, 0))
+
+    # Init tensors to avoid reallocation
+    ΔY_ = similar(ΔY)
+    Y_  = similar(Y)
+
+    num_downsamp = length(L.C)
+    for j=num_downsamp:-1:1
+        if set_grad
+            ΔY_, Y_ = L.C[j].forward((ΔY, Y))
+        else
+            ΔY_, Δθ_C1, Y_  = L.C[j].forward((ΔY, Y); set_grad=set_grad)  
+        end
+  
+        ΔYl_, ΔYr_ = tensor_split(ΔY_)
+        Y1_,  Y2_  = tensor_split(Y_)
+  
+        if set_grad
+            ΔYl_ .= L.RB[j].backward(ΔYr_, Y1_) + ΔYl_ 
+        else
+            ΔY_RB, Δθ_RB = L.RB[j].backward(ΔYr_, Y1_; set_grad=set_grad)
+            ΔYl_ .= ΔY_RB + ΔYl_ 
+        end
+
+        Y2_ .-= L.RB[j].forward(Y1_)
+
+        tensor_cat!(ΔY_, ΔYl_, ΔYr_)
+        tensor_cat!(Y_, Y1_,  Y2_)
+
+        if set_grad
+            ΔY, Y = L.C[j].inverse((ΔY_, Y_))
+        else
+            ΔY, Δθ_C2, Y = L.C[j].inverse((ΔY_, Y_); set_grad=set_grad)  
+            p1 = cat(p1, Δθ_C1+Δθ_C2; dims=1)
+            p2 = cat(p2, Δθ_RB; dims=1)
+        end
     end
     
-    set_grad ? (return ΔX, X) : (return ΔX, cat(Δθ_C1+Δθ_C2, Δθ_RB; dims=1), X)
+    set_grad ? (return ΔY, Y) : (ΔY, cat(p1, p2; dims=1), Y)
```

Comment:
Would be good to keep the naming convention the same, i.e., this should return `ΔX`.